io_scene_godot/converters/material_node_tree/shader_functions.py

```diff
@@ -0,0 +1,678 @@
+    outval = clamp(pow(val1, val2), 0.0, 1.0);
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeMath_LOGARITHM_CLAMP'] = ScriptsFunction(
+    code="""
+void math_node_log_clamp(float val1, float val2, out float outval){
+    if (val1 > 0.0  && val2 > 0.0)
+        outval = clamp(log2(val1) / log2(val2), 0.0, 1.0);
+    else
+        outval = 0.0;
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeVectorMath_ADD'] = ScriptsFunction(
+    code="""
+void math_node_vec_add(vec3 v1, vec3 v2, out vec3 outvec, out float outval) {
+  outvec = v1 + v2;
+  outval = (abs(outvec[0]) + abs(outvec[1]) + abs(outvec[2])) * 0.333333;
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeVectorMath_SUBTRACT'] = ScriptsFunction(
+    code="""
+void math_node_vec_sub(vec3 v1, vec3 v2, out vec3 outvec, out float outval) {
+  outvec = v1 - v2;
+  outval = (abs(outvec[0]) + abs(outvec[1]) + abs(outvec[2])) * 0.333333;
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeVectorMath_AVERAGE'] = ScriptsFunction(
+    code="""
+void math_node_vec_avg(vec3 v1, vec3 v2, out vec3 outvec, out float outval) {
+  outvec = v1 + v2;
+  outval = length(outvec);
+  outvec = normalize(outvec);
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeVectorMath_DOT_PRODUCT'] = ScriptsFunction(
+    code="""
+void math_node_vec_dot(vec3 v1, vec3 v2, out vec3 outvec, out float outval) {
+  outvec = vec3(0);
+  outval = dot(v1, v2);
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeVectorMath_CROSS_PRODUCT'] = ScriptsFunction(
+    code="""
+void math_node_vec_cross(vec3 v1, vec3 v2, out vec3 outvec, out float outval) {
+  outvec = cross(v1, v2);
+  outval = length(outvec);
+  outvec /= outval;
+}
+"""
+)
+
+NODE_FUNCTIONS['ShaderNodeVectorMath_NORMALIZE'] = ScriptsFunction(
+    code="""
+void math_node_vec_normalize(vec3 v, out vec3 outvec, out float outval) {
+  outval = length(v);
+  outvec = normalize(v);
+}
+"""
+)
+
+
+COMMON_FUNCTIONS = [
+    ScriptsFunction(
+        code="""
+void fresnel(vec3 view_dir, vec3 normal, float ior, out float kr) {
+// reference [https://www.scratchapixel.com/lessons/
+// 3d-basic-rendering/introduction-to-shading/reflection-refraction-fresnel]
+    float cosi = clamp(-1.0, 1.0, dot(view_dir, normal));
+    float etai = 1.0, etat = ior;
+    if (cosi > 0.0) {
+        float tmp = etai;
+        etai = etat;
+        etat = tmp;
+    }
+    // Compute sini using Snell's law
+    float sint = etai / etat * sqrt(max(0.0, 1.0 - cosi * cosi));
+    // Total internal reflection
+    if (sint >= 1.0) {
+        kr = 1.0;
+    }
+    else {
+        float cost = sqrt(max(0.0, 1.0 - sint * sint));
+        cosi = abs(cosi);
+        float Rs = ((etat * cosi) - (etai * cost))
+                    / ((etat * cosi) + (etai * cost));
+        float Rp = ((etai * cosi) - (etat * cost))
+                    / ((etai * cosi) + (etat * cost));
+        kr = (Rs * Rs + Rp * Rp) * 0.5;
+    }
+}
+"""
+    )
+]
+
+FUNTION_LIB = {func.name: func for func in NODE_FUNCTIONS.values()}
```

Comment:
1) typo
2) Two methods for accessing the nodes? One which is (eg) `ShaderNodeVectorMath_NORMALIZE` and the other which is `math_node_vec_normalize`. I can see why you did it (so you can retrieve a single function more easily), but I can't help feeling that there may be a better way, either always using `ShaderNodeVectorMath_NORMALIZE` to retrieve a function by name, or perhaps having generate_node_identifier doing something more advanced (eg convert camelcase to snakecase and remove the "ShaderNode." I'm not sure. It may be that this is the cleanest implementation.

Reply:
The `ShaderNodeVectorMath_..` is actually a `bl_idname` of the node, I am trying to have a mapping directly from node to its function, then I can use a single python function parse many types of node.   Later I found I still need to get some function outside the parsing node procedure, so add another interface to get function by its name.

It is a good suggestion to convert `bl_idname` to snake case and use them as function name, thank you!